### PR TITLE
[Feature] Add fps op from mmdet3d

### PR DESCRIPTION
**Pull request by DCNSW**
## Motivation

Add furthest point sample cuda operation from mmdet3d (branch: v1.0.0.dev0).

## Modification

Several files in mmcv/ops folder.

## BC-breaking (Optional)

No.

## Use cases (Optional)

from mmcv.ops import furthest_point_sample, furthest_point_sample_with_dist

### CODE REVIEW

File: tests/test_ops/test_furthest_point_sample.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                         [-0.0518, 3.7251, -0.3950]]]).cuda()
+
+    idx = furthest_point_sample(xyz, 3)
+    expected_idx = torch.tensor([[0, 2, 4], [0, 2, 1]]).cuda()
+    assert torch.all(idx == expected_idx)
+
+
+def test_fps_with_dist():
+    if not torch.cuda.is_available():
+        pytest.skip()
```
Comment (zhouzaida) [inline tests/test_ops/test_furthest_point_sample.py:26]:
```suggestion
@pytest.mark.skipif(
    not torch.cuda.is_available(), reason='requires CUDA support')
def test_fps_with_dist():
```

File: mmcv/ops/points_sampler.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        self.num_point = num_point
+        self.fps_sample_range_list = fps_sample_range_list
+        self.samplers = nn.ModuleList()
+        for fps_mod in fps_mod_list:
+            self.samplers.append(get_sampler_type(fps_mod)())
+        self.fp16_enabled = False
+
+    @force_fp32()
+    def forward(self, points_xyz, features):
+        """forward.
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:97]:
```suggestion
        """forward.
```

File: mmcv/ops/points_sampler.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    """
+
+    def __init__(self):
+        super(FS_Sampler, self).__init__()
+
+    def forward(self, points, features, npoint):
+        """Sampling points with FS_Sampling."""
+        assert features is not None, \
+            'feature input to FS_Sampler should not be None'
+        features_for_fps = torch.cat([points, features.transpose(1, 2)], dim=2)
```
Comment (grimoire) [inline mmcv/ops/points_sampler.py:182]: We can create DFPS_Sampler and FFPS_Sampler instance and use their forward here.

File: mmcv/ops/points_sampler.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    length_b = point_feat_b.shape[1]
+    num_channel = point_feat_a.shape[-1]
+    # [bs, n, 1]
+    a_square = torch.sum(point_feat_a.unsqueeze(dim=2).pow(2), dim=-1)
+    # [bs, 1, m]
+    b_square = torch.sum(point_feat_b.unsqueeze(dim=1).pow(2), dim=-1)
+    a_square = a_square.repeat((1, 1, length_b))  # [bs, n, m]
+    b_square = b_square.repeat((1, length_a, 1))  # [bs, n, m]
+
+    coor = torch.matmul(point_feat_a, point_feat_b.transpose(1, 2))
```
Comment (grimoire) [inline mmcv/ops/points_sampler.py:33]: `coor` is not easy to understand, better use another name.

File: mmcv/ops/points_sampler.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        Tensor: (B, N, M) Distance between each pair points.
+    """
+    length_a = point_feat_a.shape[1]
+    length_b = point_feat_b.shape[1]
+    num_channel = point_feat_a.shape[-1]
+    # [bs, n, 1]
+    a_square = torch.sum(point_feat_a.unsqueeze(dim=2).pow(2), dim=-1)
+    # [bs, 1, m]
+    b_square = torch.sum(point_feat_b.unsqueeze(dim=1).pow(2), dim=-1)
+    a_square = a_square.repeat((1, 1, length_b))  # [bs, n, m]
```
Comment (grimoire) [inline mmcv/ops/points_sampler.py:30]: This repeat might not be necessary since binary ops can be broadcast.

File: mmcv/ops/points_sampler.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        self.fps_sample_range_list = fps_sample_range_list
+        self.samplers = nn.ModuleList()
+        for fps_mod in fps_mod_list:
+            self.samplers.append(get_sampler_type(fps_mod)())
+        self.fp16_enabled = False
+
+    @force_fp32()
+    def forward(self, points_xyz, features):
+        """forward.
+
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:94]:
```suggestion
        """
```

File: mmcv/ops/points_sampler.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        last_fps_end_index = 0
+
+        for fps_sample_range, sampler, npoint in zip(
+                self.fps_sample_range_list, self.samplers, self.num_point):
+            assert fps_sample_range < points_xyz.shape[1]
+
+            if fps_sample_range == -1:
+                sample_points_xyz = points_xyz[:, last_fps_end_index:]
+                sample_features = features[:, :, last_fps_end_index:] if \
+                    features is not None else None
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:112]:
```suggestion
                if features is not None:
                    sample_features = features[:, :, last_fps_end_index:]
                else:
                    sample_features = None
```

File: mmcv/ops/points_sampler.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            else:
+                sample_points_xyz = \
+                    points_xyz[:, last_fps_end_index:fps_sample_range]
+                sample_features = \
+                    features[:, :, last_fps_end_index:fps_sample_range] if \
+                    features is not None else None
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:118]:
```suggestion
                    if features is not None:
                        sample_features = features[:, :, last_fps_end_index:fps_sample_range]
                    else:
                        sample_features = None
```

File: mmcv/ops/points_sampler.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+
+class DFPS_Sampler(nn.Module):
+    """DFPS_Sampling.
+
+    Using Euclidean distances of points for FPS.
+    """
+
+    def __init__(self):
+        super(DFPS_Sampler, self).__init__()
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:137]:
```suggestion
        super().__init__()
```

File: mmcv/ops/points_sampler.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+            FS: using F-FPS and D-FPS simultaneously.
+        fps_sample_range_list (list[int], optional):
+            Range of points to apply FPS. Default: [-1].
+    """
+
+    def __init__(self,
+                 num_point: List[int],
+                 fps_mod_list: List[str] = ['D-FPS'],
+                 fps_sample_range_list: List[int] = [-1]):
+        super(Points_Sampler, self).__init__()
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:78]:
```suggestion
        super().__init__()
```

File: mmcv/ops/points_sampler.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+
+class FFPS_Sampler(nn.Module):
+    """FFPS_Sampler.
+
+    Using feature distances for FPS.
+    """
+
+    def __init__(self):
+        super(FFPS_Sampler, self).__init__()
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:152]:
```suggestion
        super().__init__()
```

File: mmcv/ops/points_sampler.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+
+class FS_Sampler(nn.Module):
+    """FS_Sampling.
+
+    Using F-FPS and D-FPS simultaneously.
+    """
+
+    def __init__(self):
+        super(FS_Sampler, self).__init__()
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:172]:
```suggestion
        super().__init__()
```

File: mmcv/ops/points_sampler.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    """
+    if sampler_type == 'D-FPS':
+        sampler = DFPS_Sampler
+    elif sampler_type == 'F-FPS':
+        sampler = FFPS_Sampler
+    elif sampler_type == 'FS':
+        sampler = FS_Sampler
+    else:
+        raise ValueError('Only "sampler_type" of "D-FPS", "F-FPS", or "FS"'
+                         f' are supported, got {sampler_type}')
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:55]:
```suggestion
    sampler_mappings = {
        'D-FPS': DFPS_Sampler,
        'F-FPS': FFPS_Sampler,
        'FS': FS_Sampler,
    }
    try:
        return sampler_mappings[sampler_type]
    except KeyError:
        raise KeyError(f'Supported `sampler_type` is {sampler_mappings.keys()}, but got {sampler_type}')
```

File: mmcv/ops/points_sampler.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+    corr_matrix = torch.matmul(point_feat_a, point_feat_b.transpose(1, 2))
+
+    dist = a_square + b_square - 2 * corr_matrix
+    if norm:
+        dist = torch.sqrt(dist) / num_channel
+    return dist
+
+
+def get_sampler_type(sampler_type):
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:37]:
```suggestion
def get_sampler_cls(sampler_type):
```

File: mmcv/ops/points_sampler.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        # FPS would be applied to different fps_mod in the list,
+        # so the length of the num_point should be equal to
+        # fps_mod_list and fps_sample_range_list.
+        assert len(num_point) == len(fps_mod_list) == len(
+            fps_sample_range_list)
+        self.num_point = num_point
+        self.fps_sample_range_list = fps_sample_range_list
+        self.samplers = nn.ModuleList()
+        for fps_mod in fps_mod_list:
+            self.samplers.append(get_sampler_type(fps_mod)())
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:88]:
```suggestion
            self.samplers.append(get_sampler_cls(fps_mod)())
```

File: mmcv/ops/points_sampler.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+    elif sampler_type == 'FS':
+        sampler = FS_Sampler
+    else:
+        raise ValueError('Only "sampler_type" of "D-FPS", "F-FPS", or "FS"'
+                         f' are supported, got {sampler_type}')
+
+    return sampler
+
+
+class Points_Sampler(nn.Module):
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:60]: should we rename the `Points_Sampler` to `PointsSampler`

File: mmcv/ops/points_sampler.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        indices = torch.cat(indices, dim=1)
+
+        return indices
+
+
+class DFPS_Sampler(nn.Module):
+    """DFPS_Sampling.
+
+    Using Euclidean distances of points for FPS.
+    """
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:134]:
```suggestion
    """Use Euclidean distances of points for FPS."""
```

File: mmcv/ops/points_sampler.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+        """Sampling points with D-FPS."""
+        fps_idx = furthest_point_sample(points.contiguous(), npoint)
+        return fps_idx
+
+
+class FFPS_Sampler(nn.Module):
+    """FFPS_Sampler.
+
+    Using feature distances for FPS.
+    """
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:149]:
```suggestion
    """Using feature distances for FPS."""
```

File: mmcv/ops/points_sampler.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+            features_for_fps, features_for_fps, norm=False)
+        fps_idx = furthest_point_sample_with_dist(features_dist, npoint)
+        return fps_idx
+
+
+class FS_Sampler(nn.Module):
+    """FS_Sampling.
+
+    Using F-FPS and D-FPS simultaneously.
+    """
```
Comment (zhouzaida) [inline mmcv/ops/points_sampler.py:169]:
```suggestion
    """Use F-FPS and D-FPS simultaneously."""
```

File: mmcv/ops/points_sampler.py (hunk 16 above)
Comment (DCNSW) [reply mmcv/ops/points_sampler.py:60]: Since our main goal is to be consistent with mmdet3d, maybe we should keep the original name?  @ZwwWayne

Comment (ZwwWayne) [reply mmcv/ops/points_sampler.py:60]: Use PointsSampler

**Review by zhouzaida — APPROVED** [review]
LGTM

**Review by grimoire — APPROVED** [review]
LGTM

**Pull request merged**